h2o-persist-hdfs/src/main/java/water/persist/S3ATokenRefresherFactory.java

```diff
@@ -0,0 +1,79 @@
+package water.persist;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import water.H2O;
+import water.Paxos;
+import water.persist.security.HdfsDelegationTokenRefresher;
+import water.util.Log;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+public class S3ATokenRefresherFactory {
+
+    private static final String H2O_DYNAMIC_AUTH_S3A_TOKEN_REFRESHER_ENABLED = "h2o.auth.dynamicS3ATokenRefresher.enabled";
+
+    private final Configuration conf;
+    private final String tmpDir;
+    
+    private final Set<String> bucketsWithDelegationToken = Collections.synchronizedSet(new HashSet<>());
+    private final Object GENERATION_LOCK = new Object();
+
+    S3ATokenRefresherFactory(Configuration conf, String tmpDir) {
+        this.conf = conf;
+        this.tmpDir = tmpDir;
+    }
+
+    boolean startDelegationTokenRefresher(Path p) throws IOException {
+        if (Paxos._cloudLocked && H2O.CLOUD.leader() != H2O.SELF) {
+            // fast path - cloud already locked, and I am not the leader, give up - only the cloud leader is allowed to refresh the tokens
+            return false;
+        }
+
+        final URI uri = p.toUri();
+        if (!"s3a".equalsIgnoreCase(uri.getScheme())) {
+            // only S3A needs to generate delegation token
+            if (Log.isLoggingFor(Log.DEBUG)) {
+                Log.debug("Delegation token refresh is only needed for s3a, requested URI: " + uri);
+            }
+            return false;
+        }
+
+        // Important make sure the cloud is locked in order to guarantee that the leader will distribute credentials
+        // to all nodes and don't do refresh only for itself (which can happen if cloud is not yet locked)
+        Paxos.lockCloud("S3A Token Refresh");
+
+        if (H2O.CLOUD.leader() != H2O.SELF) {
+            // we are not a leader node in a locked cloud, give up
+            return true;
```

Comment:
It's not clear to me - why this returns `true`? If I am not a leader node and I've locked the cloud, then it means that the leader haven't been here and the token is not generated yet, because the leader would have locked the cloud already, otherwise I would return false on L34. So I return `true` but I don't have the token. And in getFileSystem I could potentially break from `for` cycle and do `FileSystem.get(path.toUri(), CONF);` even though I don't have the token, isn't it?

Reply:
The semantics of the method is probably not ideal: the result means - was the `path` handled or do we need to keep trying other "refresher factories"?

Line `34` returns false because we don't know yet if given `path` is an `S3A` path or something else. That is why it returns false - keep trying other refreshers.

On line `52` we already know we are dealing with `S3A` and we are not the leader. We are therefore saying the `path` was handled (by not doing anything because we are not leader).

> So I return true but I don't have the token. And in getFileSystem I could potentially break from for cycle and do FileSystem.get(path.toUri(), CONF); even though I don't have the token, isn't it?

Yes, because it is built on the assumption that the leader is handling everything and distributing the tokens. Every interaction with the filesystem starts on the leader node (that is the one called by the user), the leader then can submit tasks that do access the filesystem on non-leader nodes but these nodes already assume the token exists.

Reply:
I see now, thanks for explanation 👌

Reply:
Thanks for the comment @koniecsveta - I added javadoc and explained the true/false at each exit point of the method